<launch>
    <!-- ****** System Arguments ****** -->
    <!-- Live or reading a rosbag file (offline) -->
    <arg name="offline" default="true" />
     <!-- Flag to launch rviz -->
    <arg name="launch_rviz" default="true" />
    <!-- Check to see if loads ORB-SLAM's TF values or not -->
    <param name="use_sim_time" value="$(arg offline)" />

    <!-- Main node -->
    <node name="orb_slam3" pkg="orb_slam3_ros" cwd="node" type="ros_stereo_inertial" output="screen">
        <!-- Remap to prepare the raw images -->
        <remap from="/camera/right/image_raw"   to="/camera/infra1/image_rect_raw"/>
        <remap from="/camera/left/image_raw"    to="/camera/infra2/image_rect_raw"/>

        <!-- Remap to prepare the inertial data -->
        <remap from="/imu"                      to="/camera/imu"/>

        <!-- Parameters for loading setting files -->
        <param name="voc_file"        type="string" value="$(find orb_slam3_ros)/orb_slam3/Vocabulary/ORBvoc.txt.bin"/>
        <param name="settings_file"   type="string" value="$(find orb_slam3_ros)/config/RGB-D/LuxVSD_RS_D435i_640.yaml"/>
        <param name="sys_params_file" type="string" value="$(find orb_slam3_ros)/config/common_system_params.yaml"/>

        <!-- Parameters for ROS -->
        <param name="roll" type="double" value="0" />
        <param name="yaw" type="double" value="1.5697" />
        <param name="pitch" type="double" value="-1.5697" />
        <param name="frame_map"    type="string"   value="map" />
        <param name="imu_frame_id"      type="string"   value="imu" />
        <param name="enable_pangolin"   type="bool"     value="false" />
        <param name="world_frame_id"    type="string"   value="world" />
        <param name="cam_frame_id"      type="string"   value="camera" />
    </node>

    <!-- Transformation parameters -->
    <node pkg="tf" type="static_transform_publisher" name="bc_to_se" args="0 -3 0 0 0 0 plane room 10" />
    <node pkg="tf" type="static_transform_publisher" name="world_to_bc" args="0 -3 0 0 0 0 world plane 10" />
    <node pkg="tf" type="static_transform_publisher" name="camera_to_camera_optical" args="0 0 0 0 0 0 camera camera_color_optical_frame 10" />
    
    <!-- Visualization -->
    <node if="$(arg launch_rviz)" name="rviz" pkg="rviz" type="rviz" args="-d $(find orb_slam3_ros)/config/Visualization/vsgraphs_stereo_imu.rviz" output="screen" />
    
    <!-- Trajectory path -->
    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_orb_slam3" output="screen" ns="orb_slam3_ros" >
        <param name="/source_frame_name" value="/imu" />
        <param name="/target_frame_name" value="/map" />
        <param name="/trajectory_update_rate" value="20.0" />
        <param name="/trajectory_publish_rate" value="20.0" />
    </node>
</launch>